templates/consul.sysv.erb

```diff
@@ -61,18 +61,42 @@ start() {
 }
 
 stop() {
+        DELAY=5 # seconds maximum to wait for a leave
+
         echo -n "Shutting down consul: "
+        mkpidfile
+
         # If consul is not acting as a server, exit gracefully
+        # Use SIGINT to create a "leave" event, unless the user has explicitly
+        # changed that behavior in the Consul config.
         if ("${CONSUL}" info ${RPC_ADDR} 2>/dev/null | grep -q 'server = false' 2>/dev/null) ; then
-            "$CONSUL" leave ${RPC_ADDR}
-        fi
+            consul_pid=$(cat $PID_FILE)
+            killproc $KILLPROC_OPT $CONSUL -INT
+            retcode=$?
 
-        # If acting as a server, or if leave failed, kill it.
-        mkpidfile
-        killproc $KILLPROC_OPT $CONSUL -9
+            # We'll wait if necessary to make sure the leave works, and return
+            # early if we can.  If not, escalate to harsher signals.
+            try=0
+            while [ $try -lt $DELAY ]; do
+                if ! checkpid $consul_pid ; then
+                    rm -f /var/lock/subsys/consul
+                    return $retcode
+                fi
+                sleep 1
+                let try+=1
+            done
+        fi
 
+        # If acting as a server, use a SIGTERM to avoid a leave.
+        # This behavior is also configurable.  Avoid doing a "leave" because
+        # having servers missing is a bad thing that we want to notice.
+        #
+        # A SIGTERM will mark the node as "failed" until it rejoins.
+        # killproc with no arguments uses TERM, then escalates to KILL.
+        killproc $KILLPROC_OPT $CONSUL
```

Comment:
I realize you mentioned this in the comment but can we add `-TERM` anyway just to be explicit?